File: backend/src/main/java/com/mysite/knitly/domain/community/post/entity/Post.java
package com.mysite.knitly.domain.community.post.entity;

import com.mysite.knitly.domain.community.comment.entity.Comment;
import com.mysite.knitly.domain.user.entity.User;
import com.mysite.knitly.global.jpa.BaseTimeEntity;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Where;
import java.util.ArrayList;
import java.util.List;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
@Entity
@Table(name = "posts")
@Where(clause = "is_deleted = false")
public class Post extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "post_id")
    private Long id;

    @Column(nullable = false, length = 100)
    private String title;

    @Lob
    @Column(nullable = false)
    private String content;

    // 다중 이미지 URL
    @Builder.Default
    @ElementCollection
    @CollectionTable(name = "post_images", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "url", nullable = false, length = 512)
    @OrderColumn(name = "sort_order")
    private List<String> imageUrls = new ArrayList<>();

    // 현재 썸네일(0번 이미지)을 반환 (없으면 null)
    public String getThumbnail() {
        return (imageUrls == null || imageUrls.isEmpty()) ? null : imageUrls.get(0);
    }

    // 썸네일을 지정 인덱스 이미지로 변경 (맨 앞으로 이동)
    public void promoteThumbnail(int index) {
        if (imageUrls == null) return;
        if (index <= 0 || index >= imageUrls.size()) return;
        String picked = imageUrls.remove(index);
        imageUrls.add(0, picked);
    }

    // 썸네일을 지정 URL 이미지로 변경
    public void promoteThumbnailByUrl(String url) {
        if (imageUrls == null || url == null) return;
        int idx = imageUrls.indexOf(url);
        if (idx > 0) promoteThumbnail(idx);
    }

    @Enumerated(EnumType.STRING)
    @Column(name = "post_category", nullable = false, columnDefinition = "ENUM('FREE','QUESTION','TIP')")
    private PostCategory category;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User author;

    @Builder.Default
    @Column(name = "is_deleted", nullable = false)
    private boolean deleted = false;

    @Builder.Default
    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Comment> comments = new ArrayList<>();

    public void addComment(Comment comment) {
        comments.add(comment);
        comment.setPost(this);
    }

    public void softDelete() { this.deleted = true; }

    // 이미지 수정
    public void update(String title, String content, PostCategory category) {
        this.title = title;
        this.content = content;
        this.category = category;
    }

    // 이미지 교체
    public void replaceImages(List<String> newUrls) {
        this.imageUrls.clear();
        if (newUrls != null) {
            this.imageUrls.addAll(newUrls);
        }
    }

        public boolean isAuthor(User user) {
        return user != null && author != null && author.getUserId().equals(user.getUserId());
    }
}

File: backend/src/main/kotlin/com/mysite/knitly/domain/community/post/entity/Post.kt
package com.mysite.knitly.domain.community.post.entity

import com.mysite.knitly.domain.community.comment.entity.Comment
import com.mysite.knitly.domain.user.entity.User
import com.mysite.knitly.global.jpa.BaseTimeEntity
import jakarta.persistence.*
import org.hibernate.annotations.Where

@Entity
@Table(name = "posts")
@Where(clause = "is_deleted = false")
class Post(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "post_id")
    val id: Long? = null, // PK는 생성 후 변경되지 않음

    @Column(nullable = false, length = 100)
    var title: String,

    @Lob
    @Column(nullable = false)
    var content: String,

    @ElementCollection
    @CollectionTable(
        name = "post_images",
        joinColumns = [JoinColumn(name = "post_id")]
    )
    @Column(name = "url", nullable = false, length = 512)
    @OrderColumn(name = "sort_order")
    val imageUrls: MutableList<String> = mutableListOf(),

    @Enumerated(EnumType.STRING)
    @Column(
        name = "post_category",
        nullable = false,
        columnDefinition = "ENUM('FREE','QUESTION','TIP')"
    )
    var category: PostCategory,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    val author: User, // 작성자는 변경되지 않음

    @Column(name = "is_deleted", nullable = false)
    var deleted: Boolean = false,

    @OneToMany(
        mappedBy = "post",
        cascade = [CascadeType.ALL],
        orphanRemoval = true
    )
    val comments: MutableList<Comment> = mutableListOf()

) : BaseTimeEntity() {

    fun getThumbnail(): String? {
        return if (imageUrls.isEmpty()) null else imageUrls[0]
    }

    fun promoteThumbnail(index: Int) {
        if (index <= 0 || index >= imageUrls.size) return
        val picked = imageUrls.removeAt(index)
        imageUrls.add(0, picked)
    }

    fun promoteThumbnailByUrl(url: String?) {
        if (url == null) return
        val idx = imageUrls.indexOf(url)
        if (idx > 0) promoteThumbnail(idx)
    }

    fun addComment(comment: Comment) {
        comments.add(comment)
    }

    fun softDelete() {
        deleted = true
    }

    fun update(title: String, content: String, category: PostCategory) {
        this.title = title
        this.content = content
        this.category = category
    }

    fun replaceImages(newUrls: List<String>?) {
        imageUrls.clear()
        if (newUrls != null) {
            imageUrls.addAll(newUrls)
        }
    }

    fun isAuthor(user: User?): Boolean {
        return user != null && author.userId == user.userId
    }
}
